x-pack/plugins/observability/server/services/slo/resource_installer.ts

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type {
+  ClusterPutComponentTemplateRequest,
+  IndicesPutIndexTemplateRequest,
+  IngestPutPipelineRequest,
+} from '@elastic/elasticsearch/lib/api/typesWithBodyKey';
+import type { ElasticsearchClient, Logger } from '@kbn/core/server';
+
+import {
+  SLO_COMPONENT_TEMPLATE_MAPPINGS_NAME,
+  SLO_COMPONENT_TEMPLATE_SETTINGS_NAME,
+  SLO_INDEX_TEMPLATE_NAME,
+  SLO_INGEST_PIPELINE_NAME,
+} from '../../assets/constants';
+import { sloMappingsTemplate } from '../../assets/component_templates/slo_mappings_template';
+import { sloSettingsTemplate } from '../../assets/component_templates/slo_settings_template';
+import { getSLOIndexTemplate } from '../../assets/index_templates/slo_index_templates';
+import { getSLOPipelineTemplate } from '../../assets/ingest_templates/slo_pipeline_template';
+
+export class ResourceInstaller {
+  constructor(private esClient: ElasticsearchClient, private logger: Logger) {}
+
+  public async ensureCommonResourcesInstalled(): Promise<void> {
+    const alreadyInstalled = await this.areResourcesAlreadyInstalled();
+
+    if (alreadyInstalled) {
+      this.logger.debug(
+        `Skipping installation of resources shared for SLO since they already exist`
+      );
+      return;
+    }
+
+    try {
+      await Promise.all([
+        this.createOrUpdateComponentTemplate({
+          name: SLO_COMPONENT_TEMPLATE_MAPPINGS_NAME,
+          ...sloMappingsTemplate,
+        }),
+
+        this.createOrUpdateComponentTemplate({
+          name: SLO_COMPONENT_TEMPLATE_SETTINGS_NAME,
+          ...sloSettingsTemplate,
+        }),
+      ]);
+
+      await this.createOrUpdateIndexTemplate({
+        name: SLO_INDEX_TEMPLATE_NAME,
+        ...getSLOIndexTemplate(`${SLO_INDEX_TEMPLATE_NAME}-*`, [
+          SLO_COMPONENT_TEMPLATE_MAPPINGS_NAME,
+          SLO_COMPONENT_TEMPLATE_SETTINGS_NAME,
+        ]),
+      });
+
+      await this.createOrUpdateIngestPipelineTemplate({
+        id: SLO_INGEST_PIPELINE_NAME,
+        ...getSLOPipelineTemplate(`${SLO_INDEX_TEMPLATE_NAME}-default-`),
+      });
+    } catch (err) {
+      this.logger.error(`Error installing resources shared for SLO - ${err.message}`);
+      throw err;
+    }
+  }
+
+  private async areResourcesAlreadyInstalled(): Promise<boolean> {
+    const indexTemplateExists = await this.esClient.indices.existsIndexTemplate({
+      name: SLO_INDEX_TEMPLATE_NAME,
+    });
+
+    let ingestPipelineExists = false;
+    try {
+      await this.esClient.ingest.getPipeline({
+        id: SLO_INGEST_PIPELINE_NAME,
+      });
+
+      ingestPipelineExists = true;
```

Comment:
Later, we might want to add some checks on the `_meta.version` for example instead of just checking the existence of the ingest pipeline

Reply:
I wonder if we shouldn't put a version on the mappings, pipeline, settings, and in the index pattern name in the pipeline so that when/if we change the mappings we can figure out what to do with the existing indices.

Reply:
We'll need to figure out the migration logic later. I'm going to create an issue for that. 
For the time being, I'll just add a version into each templates, and compare the ingest pipeline version when checking for the existence.

Reply:
Issue created for tracking: https://github.com/elastic/kibana/issues/139931